src/cpp/src/sequence_group.hpp

```diff
@@ -396,7 +421,17 @@ class SequenceGroup {
     }
 
     size_t get_num_logical_blocks() const {
-        return (get_context_len() + m_block_size - 1) / m_block_size;
+        size_t max_num_logical_blocks = 0;
+        for (const auto& seq : m_sequences) {
+            max_num_logical_blocks = std::max(max_num_logical_blocks,
+                                              (get_context_len() - seq->get_num_evicted_tokens() + m_block_size - 1) / m_block_size);
+        }
+        return max_num_logical_blocks;
+    }
+
+
+    size_t get_num_logical_blocks(size_t seq_id) const {
+        return (get_context_len() - get_sequence_by_id(seq_id)->get_num_evicted_tokens() + m_block_size - 1) / m_block_size;
```

Comment:
all sequence must have the same number of block, must not they?

Reply:
Refactored to presume same number of blocks in sequences in a group